src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,69 @@
+package duke;
+
+import command.Command;
+import task.TaskList;
+import ui.UserInterface;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.text.ParseException;
+
+/**
+ * Runs the program skeleton: Event flow.
+ */
+public class Duke {
+
+    private Storage storage;
+    private TaskList tasks;
+    private UserInterface ui;
+
+    public Duke(String filepath) {
+        startup(filepath);
+    }
+
+    /**
+     * Parses user inputs and return appropriate replies.
+     * @param input User's action.
+     * @return Reply from duke.
+     */
+    public String getResponse(String input) {
+        try {
+            boolean isExit = false;
+
+            Command c = Parser.parse(input);
+            c.execute(tasks, ui, storage);
+            isExit = c.isExit();
+            storage.save(tasks.getAllTasks());
+
+            if(isExit) {
+                return ui.goodbye();
+            } else {
+                return c.toString();
+            }
+
+        } catch (DukeException | ParseException | IOException e) {
```

Comment:
Would it be clearer to throw various DukeExceptions (and create more appropriate subclasses) instead of throwing DukeExceptions and various Java exceptions? This way, the user knows that the error is coming from Duke and not somewhere else. Then, you can also customise your error messages for the issue. E.g. "Save file corrupted" instead of an IndexOutOfBounds exception.

Alternatively, there's also the lazy way of using one DukeException class and passing the appropriate messages to it. That's what I did.